mempool/src/pool.rs

```diff
@@ -0,0 +1,815 @@
+use std::cmp::Ord;
+use std::cmp::Ordering;
+use std::collections::BTreeSet;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use parity_scale_codec::Encode;
+use thiserror::Error;
+
+use common::chain::transaction::Transaction;
+use common::chain::transaction::TxInput;
+use common::chain::OutPoint;
+use common::primitives::amount::Amount;
+use common::primitives::Id;
+use common::primitives::Idable;
+use common::primitives::H256;
+// TODO Add a newtype macro
+
+// TODO this willbe defined elsewhere
+const MAX_BLOCK_SIZE: usize = 1_000_000;
+
+pub const MEMPOOL_MAX_TXS: usize = 1_000_000;
+
+#[derive(PartialEq, Eq, Clone, Hash)]
+struct TxMemPoolEntry {
+    tx: Arc<Transaction>,
+    parents: BTreeSet<TxMemPoolEntry>,
+}
+
+// TODO consider mocking the fees more accurately by referencing here a global mock Chain State
+// object
+fn compare_fees(this: &TxMemPoolEntry, other: &TxMemPoolEntry) -> Ordering {
+    // We want the comparison to be such that a collection is sorted if its elements are ordered in
+    // descending order
+    other.get_fee().cmp(&this.get_fee())
+}
+
+impl TxMemPoolEntry {
+    //TODO this should really be sum of inputs minus sum of outputs
+    //But we don't yet have a way of summing of the inputs
+    //TODO handle the error bettter
+    fn get_fee(&self) -> Amount {
+        self.tx
+            .get_outputs()
+            .iter()
+            .map(|output| output.get_value())
+            .sum::<Option<_>>()
+            .expect("fee overflow")
```

Comment:
Actually we're using uint128 for amount, which will make it highly unlikely for any overflow to occur. Bitcoin uses int64 and still doesn't overflow with the whole 21 million sats. However, using expect here is wrong because a node can be attacked by sending something that overflows and crashing the node. So this should either return an error, or should be restricted to being done only after validating a transaction.

Reply:
We have a few more than 21 million though, and we support millisat-esque units like LN so we have slightly larger values than btc (in theory at least). That said it's still unlikely to overflow.

I'd return an error here

Reply:
Unlikely, unless someone attacks the node, just like they did with bitcoin:

https://thebitcoinnews.com/bitcoin-history-part-10-the-184-billion-btc-bug/

Reply:
This necessitates some reworking. Right now we are using the fee calculation as the criterion for ordering entries inside the BTreeMap/BTreeSet. This means get_fee is called not directly by us but by the code comparing the entries. The signature of `Ord` trait doesn't allow for errors. So I need to rethink how we will handle this. Suggestions welcome.

Reply:
Reworking done, made the fee a member of `TxMempoolEntry` struct so it doesn't have to be computed when comparing entries.